compiler/src/dotty/tools/repl/Rendering.scala

```diff
@@ -55,11 +55,29 @@ private[repl] class Rendering(parentClassLoader: Option[ClassLoader] = None) {
         // We need to use the ScalaRunTime class coming from the scala-library
         // on the user classpath, and not the one available in the current
         // classloader, so we use reflection instead of simply calling
-        // `ScalaRunTime.replStringOf`.
+        // `ScalaRunTime.replStringOf`. Probe for new API without extraneous newlines.
+        // For old API, try to clean up extraneous newlines by stripping suffix and maybe prefix newline.
         val scalaRuntime = Class.forName("scala.runtime.ScalaRunTime", true, myClassLoader)
-        val meth = scalaRuntime.getMethod("replStringOf", classOf[Object], classOf[Int])
-
-        (value: Object) => meth.invoke(null, value, Integer.valueOf(MaxStringElements)).asInstanceOf[String]
+        try {
+          val meth = scalaRuntime.getMethod("replStringOf", classOf[Object], classOf[Int], classOf[Boolean])
+          val truly = java.lang.Boolean.TRUE
+
+          (value: Object) => meth.invoke(null, value, Integer.valueOf(MaxStringElements), truly).asInstanceOf[String]
+        } catch {
+          case _: NoSuchMethodException =>
+            val meth = scalaRuntime.getMethod("replStringOf", classOf[Object], classOf[Int])
+
+            (value: Object) => {
+              val res = meth.invoke(null, value, Integer.valueOf(MaxStringElements)).asInstanceOf[String]
+              val len = res.length()
+              if len == 0 || res.charAt(len-1) != '\n' then
+                res
+              else if len == 1 || res.charAt(0) != '\n' then
+                res.substring(0, len-1)
```

Comment:
This code seems yo assume that there always is a newline at the end of res? Why not check explicitly if also the last character is a newline to be cut, without just assuming that its there (to be more robust and not living on assumptions...)? Or am I missing something?

Reply:
L73 checks last, L75 checks first char. I'll look again when awake.

Reply:
Yes, my bad. I was not going through all the cases in my head, so it seems to work.

Reply:
So the logic is:
`if` non-zero *or* no trailing \n `then` leave it as is
`else if` only one char *or* no leading \n `then` cut last (because we now know that there is a trailing \n)
`else` cut both first and last (because we now know that there must be both leading and trailing \n)

Reply:
Somehow I think this is easier to grok: (but maybe it's just me)
```scala
def trimNewlinesAtStartOrEnd(res: String) =
  val len = res.length
  if len == 0 || res == "\n" then ""
  else
    val from  = if res.charAt(0) == '\n' then 1 else 0
    val until = if res.charAt(len - 1) == '\n' then len - 1 else len
    res.substring(from, until)  
``` 
Take whatever version you like best :)
Simpler to just take the special cases first IMHO.
I just had to test it :)
```
scala> trimNewlinesAtStartOrEnd("")
val res0: String = ""

scala> trimNewlinesAtStartOrEnd("\n")                                                                                     
val res1: String = ""

scala> trimNewlinesAtStartOrEnd("\n\n")                                                                               
val res2: String = ""

scala> trimNewlinesAtStartOrEnd("\nhello\n")                                                                                   
val res3: String = hello

scala> trimNewlinesAtStartOrEnd("\n\n\n")                                                                                   
val res4: String = "
"
```

Reply:
BTW: Where do these surrounding newlines come from and why do they need to be trimmed away?

Reply:
It appends always for Scala 2 REPL; other question is why doesn't Scala 3 use `stringOf`. I suppose it prepends on multiline so the lines are all left-adjusted.

My time box expired on this tweak, so I didn't add a special test; the repl tests break easily. I agree those lines tax the short-term memory of the reviewer; if the Scala 2 PR is merged, they will be unused code.

Reply:
BTW `stringOf` is the original method name, not an internal helper; `replStringOf` was split out a decade ago out of fastidiousness. So my better option here was just to use `stringOf` directly. I'll take a few minutes to try that out, after I knock off of my day job.

Reply:
Ok thanks for info. LGTM.